### PR TITLE
tests: Collect tests only from the `test/` directory

### PR DESCRIPTION
**Pull request by math-fehr**
Additionally, make errors fatal in pytest.
This will make sure for instance that deprecated functions are not called with pytest

### CODE REVIEW

File: setup.cfg [hunk 1]
```diff
@@ -9,7 +9,7 @@ ignore = F403,E226,E731,E275,W503,F405,E722,E741,W504,W605,E402
 exclude = .github
 
 [tool:pytest]
-python_files = *_test.py test_*.py *_example.py example_*.py
+python_files = tests/*test_*.py
```
Comment (webmiche) [inline setup.cfg:12]:
```suggestion
python_files = tests/**/test_*.py
```

Doesn't something like this cover all potential depths of folder hierarchies?

**Comment by georgebisbas** [comment]
> This will make sure for instance that deprecated functions are not called with pytest
What is the current problem with this?

File: setup.cfg (hunk 1 above)
Comment (math-fehr) [reply setup.cfg:12]: My understanding is that this configuration file use a different syntax.
I tried exactly what you propose before, and it ended up only taking tests that have at least one subfolder. My understanding is that it consider `*` as a placeholder for any characters, *even* `/` (which is really weird, but that's my understanding.

**Comment by math-fehr** [comment]
> > This will make sure for instance that deprecated functions are not called with pytest
> > What is the current problem with this?

Well, now that we deprecate functions, we should make sure to not use them in our own codebase. The deprecation of function is meant for users using xDSL, while we should use this correctly.

**Comment by math-fehr** [comment]
I added a better regex for test classes and functions for pytest collection.
We now make clear what tests we collect.
We collect functions starting with `test_`, and classes starting with `Test_`.

**Comment by math-fehr** [comment]
Pytest was complaining that we weren't closing files, so I fixed this as well in xdsl-opt

File: setup.cfg (hunk 1 above)
Comment (webmiche) [reply setup.cfg:12]: Okay, weird stuff :sweat_smile:

**Comment by webmiche** [comment]
>This will make sure for instance that deprecated functions are not called with pytest

Which part of the PR does that? Just the -W error?

File: xdsl/xdsl_opt_main.py [hunk 2]
```diff
@@ -315,8 +321,8 @@ def print_to_output_stream(self, contents: str):
         if self.args.output_file is None:
             print(contents)
         else:
-            output_stream = open(self.args.output_file, 'w')
-            output_stream.write(contents)
+            with open(self.args.output_file, 'w') as output_stream:
```
Comment (superlopuh) [inline xdsl/xdsl_opt_main.py:324]: 👍

**Review by superlopuh — APPROVED** [review]

**Review by georgebisbas — APPROVED** [review]

**Comment by math-fehr** [comment]
> > This will make sure for instance that deprecated functions are not called with pytest
> 
> Which part of the PR does that? Just the -W error?

Yes

**Comment by superlopuh** [comment]
Oh wait this means we're no longer testing toy

**Pull request merged**